sw/device/lib/runtime/pmp.c

```diff
@@ -0,0 +1,539 @@
+    return kPmpRegionConfigureError;
+  }
+
+  if (!pmp_cfg_csr_rw(region, kPmpCsrAccessTypeRead, &cfg_csr_current)) {
+    return kPmpRegionConfigureError;
+  }
+
+  if (cfg_csr_current != cfg_csr_new) {
+    return kPmpRegionConfigureWarlError;
+  }
+
+  return kPmpRegionConfigureOk;
+}
+
+/**
+ * Writes `address` to a pmpaddr CSRs.
+ *
+ * The corresponding pmpaddrN index N is determined by `region`.
+ *
+ * PMP address must be at least 4bytes aligned, and pmpaddr holds only bits
+ * 33:2. This means that before writing an address to a pmpaddr CSR, it must be
+ * shifted 2 bits to the right.
+ *
+ * Please see:
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6.1 Physical Memory Protection CSRs",
+ * "Figure 3.26: PMP address register format, RV32".
+ *
+ * @param region PMP region to configure and set address for.
+ * @param address PMP address to be set.
+ * @return `pmp_region_configure_result_t`.
+ */
+pmp_region_configure_result_t pmp_csr_address_write(
+    pmp_region_index_t region, pmp_region_address_t address) {
+  pmp_region_address_t address_shifted = address >> PMP_ADDRESS_SHIFT;
+  if (!pmp_addr_csr_rw(region, kPmpCsrAccessTypeWrite, &address_shifted)) {
+    return kPmpRegionConfigureError;
+  }
+
+  pmp_region_address_t addr_csr_after_write;
+  if (!pmp_addr_csr_rw(region, kPmpCsrAccessTypeRead, &addr_csr_after_write)) {
+    return kPmpRegionConfigureError;
+  }
+
+  if (address_shifted != addr_csr_after_write) {
+    return kPmpRegionConfigureWarlError;
+  }
+
+  return kPmpRegionConfigureOk;
+}
+
+// TODO - short description.
+static bool pmp_cfg_permissions_set(pmp_region_permissions_t perm,
+                                    uint32_t *bitfield) {
+  switch (perm) {
+    case kPmpRegionPermissionsNone:
+      // No access is allowed.
+      break;
+    case kPmpRegionPermissionsReadOnly:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_R, true);
+      break;
+    case kPmpRegionPermissionsExecuteOnly:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_X, true);
+      break;
+    case kPmpRegionPermissionsReadExecute:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_R, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_X, true);
+      break;
+    case kPmpRegionPermissionsReadWrite:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_R, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_W, true);
+      break;
+    case kPmpRegionPermissionsReadWriteExecute:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_R, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_W, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_X, true);
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+// TODO - short description.
+static void pmp_cfg_mode_lock_set(pmp_region_lock_t lock, uint32_t *bitfield) {
+  if (lock == kPmpRegionLockLocked) {
```

Comment:
Consider inlining the condition into `bitfield_bit32`, so that the source code is branchless.

Reply:
used ternary operator instead